docs/decisions/0002-enterprise-catalog-ida.rst

```diff
@@ -0,0 +1,64 @@
+New Enterprise Catalog IDA
+--------------------------
+
+Status
+======
+
+Draft
+
+Context
+=======
+We are interested in improving the reliability and performance of our system at large, and we have identified that the enterprise catalog api and related functionality are a central dependency that often causes issues, such as users experiencing random failures during enrollment or other workflows. Almost every feature that our users interact with depends on enterprise catalogs: Enrollment, Code Redemption, Data Sharing Consent, Data transmission to external LMS systems, and API access given to select customers.
+
+Our main goals are as follows:
+
+1. Get the response time of the enterprise customer catalog endpoints to be below 1 second for 99% of calls.
+
+2. Since many features do not hit the enterprise catalog api endpoints but perform the same calculations directly in code, we would also like to improve the performance of those code paths and drastically reduce the calls to the discovery service.
+
+What is Enterprise Catalog?
+___________________________
+
+Enterprise catalogs are essentially saved elasticsearch queries attached to an Enterprise Customer. There can be many catalogs configured for a given customer, which enables different sets of learners from the same Enterprise to be given access to different sets of courses, and allows different discounts to be configured for different sets of courses.
+
+We expose two apis that are frequently used by various enterprise features: one to retrieve all of the content metadata that would be returned by the configured catalog query, and one to determine if a given course id or program id is part of the set of content returned by the catalog query. These apis exist both as python apis and REST apis that can be hit from within our edX ecosystem and externally from customers. They are wrappers around the search/all endpoint in the discovery service. Although there is caching in place, the calls to discovery are implemented synchronously, and we've found that half the time requests do not hit the cache, and thus the most of the time is spent waiting on the discovery service to return results.
```

Comment:
Not a comment on the ADR itself, but I'm wondering how easily fixable the cache misses to the `/search/all` endpoint are. The discovery theme dealt with a similar issue for the `/courses`, `/programs`, and `/course_runs` endpoints so maybe there's more improvement there?

Reply:
it is possible, but for various reasons, we don't think the current architecture is going to scale in the long term. i am hoping that if we have this in a separate service, we will:
* prove that we can iterate faster for development/bug fixes when we own our own service, instead of depending on edx-enterprise being installed in edx-platform.
* be able to build more robust/innovative functionality around catalog curation and enterprise enrollment lifecycles, depending on product needs. currently both are pain points in the system/product
* have more visibility into a key part of the enterprise system and be able to proactively monitor and alert

Reply:
Agree that I don't think current discovery will scale and it makes to account for all of those things